### PR TITLE
Fix time rounding for Newfoundland

### LINKED ISSUE #27966: [CI] TimeZoneRoundingTests#testRoundingRandom fails
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.6+periodic/1149/console

```
   > Throwable #1: java.lang.AssertionError
   >    at __randomizedtesting.SeedInfo.seed([1E275DAB660CE5C0:B169E5B952C9109D]:0)
   >    at org.elasticsearch.common.rounding.TimeZoneRoundingTests.assertInterval(TimeZoneRoundingTests.java:589)
   >    at org.elasticsearch.common.rounding.TimeZoneRoundingTests.testRoundingRandom(TimeZoneRoundingTests.java:196)
```

This is a reproducible failure:
```
gradle :core:test \
  -Dtests.seed=1E275DAB660CE5C0 \
  -Dtests.class=org.elasticsearch.common.rounding.TimeZoneRoundingTests \
  -Dtests.method="testRoundingRandom" \
  -Dtests.security.manager=true \
  -Dtests.locale=et-EE \
  -Dtests.timezone=America/Grand_Turk
```

### PR DESCRIPTION
**Pull request by DaveCTurner**
Resolves #27966. I do love a good timezone puzzle.

I updated `testDST_END_Edgecases` because I think the test itself is wrong. In its original form it failed:

```
java.lang.AssertionError: dateBetween [2000-10-29T22:10:11.747-01:00] should round down to roundedDate
Expected: 2000-10-29T00:00:00.000Z [972777600000] 
     but:  was "2000-10-29T00:00:00.000-01:00 [972781200000]"
```

Rounding to `DAY_OF_MONTH` should round down to the _closest_ earlier local-time midnight. However, in `Atlantic/Azores` there are two midnights on 2000-10-29 because the clocks went back by an hour at 0100. The test originally expected the first of them, but the second is closer. @cbuescher WDYT?

Don't know a good label for this bit of code, so I'm just labelling this PR as 'test'.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I've changed my mind. I now think that if there are two local midnights with the same date then we should round down to the first one and consider the second one to be unrounded.

The alternative, implemented here, is bad because date histograms would end up with two buckets with the same label, one of which only includes an hour's worth of data. For "normal" autumn clock changes there's one bucket with 25 hours' with of data and I think that should be what we do here.

**Comment by DaveCTurner** [comment]
An unhappy consequence of all of this is that data from the final, repeated, 59 minutes of each problematic day in Newfoundland (e.g. 2006-10-28) will be mislabelled in a date histogram as belonging to the following day (resp. 2006-10-29) since that's the closest preceding midnight. Maybe it's just the assertion that's invalid? Gah.

File: core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java [hunk 1]
```diff
@@ -527,11 +577,15 @@ public void testDST_END_Edgecases() {
...
         // to Sunday, 29 October 2000, 00:00:00 local standard time instead
+        // which means there were two midnights that day.
 
         long midnightBeforeTransition = time("2000-10-29T00:00:00", tz);
+        long midnightOfTransition = time("2000-10-29T00:00:00-01:00");
+        assertEquals(60L * 60L * 1000L, midnightOfTransition - midnightBeforeTransition);
         long nextMidnight = time("2000-10-30T00:00:00", tz);
 
-        assertInterval(midnightBeforeTransition, nextMidnight, rounding, 25 * 60, tz);
+        assertInterval(midnightBeforeTransition, midnightOfTransition, rounding, 60, tz);
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java]: Just to clarify this: this means that when we do e.g. daily date histograms in this time zone around this time change, we would get a 24h bucket for Sat 28th, then an extra 1-hour bucket for the first hour after midnight on the 29th and then a full 24h bucket for the day after the time change? Sounds like an okay trade-off to me, with the potential disadvantage that we have a very small bucket (might be confusing in stats in graphs etc, but I guess something has to five...) and that October then 32 buckets.

File: core/src/main/java/org/elasticsearch/common/rounding/Rounding.java [hunk 2]
```diff
@@ -127,54 +127,59 @@ public byte id() {
...
-                         */
-                        long previousRounded = field.roundFloor(rounded - 1);
-                        if (rounded - previousRounded < field.getDurationField().getUnitMillis()) {
-                            rounded = previousRounded;
-                        }
-                    }
+            while (true) {
+                long rounded = field.roundFloor(utcMillis);
+                long transitionTime = timeZone.previousTransition(utcMillis);
+                if (transitionTime == utcMillis) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: I think this adds a `timeZone.previousTransition()` computation even for fixed time zone cases. I think to save computation we could return even earlier in unproblematic cases when `timeZone.isFixed()` like before.

File: core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java [hunk 3]
```diff
@@ -514,6 +514,56 @@ public void testEdgeCasesTransition() {
...
+            long timeAfterSecondMidnight = time("2006-10-29T06:00:00.000-03:30");
+            long floor = rounding.round(timeAfterSecondMidnight);
+            assertThat(floor, isDate(time("2006-10-29T00:00:00.000-03:30"), tz));
+            long ceiling = rounding.nextRoundingValue(timeAfterSecondMidnight);
+            assertThat(ceiling, isDate(time("2006-10-30T00:00:00.000-03:30"), tz));
+            assertInterval(floor, timeAfterSecondMidnight, ceiling, rounding, tz);
+        }
+
+        {
+            // Times between the two midnights should be rounded down to the first one and up to the second.
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java]: So this also add the extra 1-"ghost"-hour bucket to the time line for e.g. histograms here? Same thoughts than on the other changed test case then. Also, what happens for times in the first minute? I guess they are also rounded down to "2006-10-29T00:00:00.000-02:30"? Maybe worth to add if we keep this (or also use a random date out of the first hour).

**Review by cbuescher — COMMENTED** [review]
@DaveCTurner thanks for this great solution and yet another curious time zone edge case discovered ;-)

I had a first look at the changes in the tests and left some comments around the proposed handling of some of the problematic edge cases. Before diving into the actual rounding code I'd like to get some feedback and discuss if adding those small extra buckets is what we want to do in those cases.

File: core/src/main/java/org/elasticsearch/common/rounding/Rounding.java (hunk 2 above)
Comment (DaveCTurner) [reply core/src/main/java/org/elasticsearch/common/rounding/Rounding.java]: I'd rather have fewer branches unless the extra computation is onerous. [In this case I think it makes little difference.](https://github.com/JodaOrg/joda-time/blob/f08b3ea04843ef901f20de212ab53f1e183305c3/src/main/java/org/joda/time/tz/FixedDateTimeZone.java#L60-L70)

File: core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java (hunk 3 above)
Comment (DaveCTurner) [reply core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java]: Yes, I'm now thinking that this isn't what we want: https://github.com/elastic/elasticsearch/pull/27976#issuecomment-354881706.

File: core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java [hunk 4]
```diff
@@ -584,9 +638,7 @@ private static void assertInterval(long rounded, long nextRoundingValue, Roundin
      * @param nextRoundingValue the expected upper end of the rounding interval
      * @param rounding the rounding instance
      */
-    private static void assertInterval(long rounded, long unrounded, long nextRoundingValue, Rounding rounding,
-            DateTimeZone tz) {
-        assert rounded <= unrounded && unrounded <= nextRoundingValue;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java]: Yes, I think it does hold, but the following lines (a) assert slightly stronger properties and (b) give more useful messages if they fail.

**Comment by DaveCTurner** [comment]
```
...                                           truncate to day                  round to prev local midnight       round to first local midnight
2006-10-28T23:59:58-02:30                     2006-10-28T00:00:00-02:30        2006-10-28T00:00:00-02:30          2006-10-28T00:00:00-02:30
2006-10-28T23:59:59-02:30                     2006-10-28T00:00:00-02:30        2006-10-28T00:00:00-02:30          2006-10-28T00:00:00-02:30
2006-10-29T00:00:00-02:30 <- normal midnight  2006-10-29T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30
2006-10-29T00:00:01-02:30                     2006-10-29T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:01-02:30
...                                           ...                              ...                                ...
2006-10-29T00:00:58-02:30                     2006-10-29T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30
2006-10-29T00:00:59-02:30                     2006-10-29T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30
2006-10-28T23:01:00-03:30 <- DST change       2006-10-28T00:00:00-02:30 (1)    2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30   
2006-10-28T23:01:01-03:30                     2006-10-28T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30
...                                           ...                              ...                                ...
2006-10-28T23:59:58-03:30                     2006-10-28T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30
2006-10-28T23:59:59-03:30                     2006-10-28T00:00:00-02:30        2006-10-29T00:00:00-02:30          2006-10-29T00:00:00-02:30
2006-10-29T00:00:00-03:30 <- extra midnight   2006-10-29T00:00:00-02:30        2006-10-29T00:00:00-03:30 (2)      2006-10-29T00:00:00-02:30
2006-10-29T00:00:01-03:30                     2006-10-29T00:00:00-02:30        2006-10-29T00:00:00-03:30          2006-10-29T00:00:00-02:30
...
```

Here are the three possible implementations I can think of. The first maps times to the correct day, but is not monotonic at `(1)` (as found by the failing test). The second is what's I've implemented here and is monotonic but changes the day of some local times, and also has the 1-hour long extra bucket that I think will cause confusion in histograms. The third I think is better because it doesn't have that extra bucket, but still changes the day of some local times.

**Comment by DaveCTurner** [comment]
FWIW Postgres (9.2) does something completely wacky here:

```
# select t as raw_time, t at time zone 'UTC' as utc, date_trunc('day', t) as date_trunc_day, date_trunc('hour', t) as date_trunc_hour from times order by t;
         raw_time          |         utc         |      date_trunc_day       |      date_trunc_hour
---------------------------+---------------------+---------------------------+---------------------------
 2006-10-28 23:59:58-02:30 | 2006-10-29 02:29:58 | 2006-10-28 00:00:00-02:30 | 2006-10-28 23:00:00-02:30
 2006-10-28 23:59:59-02:30 | 2006-10-29 02:29:59 | 2006-10-28 00:00:00-02:30 | 2006-10-28 23:00:00-02:30
 2006-10-29 00:00:00-02:30 | 2006-10-29 02:30:00 | 2006-10-29 00:00:00-03:30 | 2006-10-29 00:00:00-02:30
 2006-10-29 00:00:01-02:30 | 2006-10-29 02:30:01 | 2006-10-29 00:00:00-03:30 | 2006-10-29 00:00:00-02:30
 2006-10-29 00:00:58-02:30 | 2006-10-29 02:30:58 | 2006-10-29 00:00:00-03:30 | 2006-10-29 00:00:00-02:30
 2006-10-29 00:00:59-02:30 | 2006-10-29 02:30:59 | 2006-10-29 00:00:00-03:30 | 2006-10-29 00:00:00-02:30
 2006-10-28 23:01:00-03:30 | 2006-10-29 02:31:00 | 2006-10-28 00:00:00-02:30 | 2006-10-29 00:00:00-02:30
 2006-10-28 23:01:01-03:30 | 2006-10-29 02:31:01 | 2006-10-28 00:00:00-02:30 | 2006-10-29 00:00:00-02:30
 2006-10-28 23:59:58-03:30 | 2006-10-29 03:29:58 | 2006-10-28 00:00:00-02:30 | 2006-10-29 00:00:00-02:30
 2006-10-28 23:59:59-03:30 | 2006-10-29 03:29:59 | 2006-10-28 00:00:00-02:30 | 2006-10-29 00:00:00-02:30
 2006-10-29 00:00:00-03:30 | 2006-10-29 03:30:00 | 2006-10-29 00:00:00-03:30 | 2006-10-29 00:00:00-03:30
 2006-10-29 00:00:01-03:30 | 2006-10-29 03:30:01 | 2006-10-29 00:00:00-03:30 | 2006-10-29 00:00:00-03:30
(12 rows)
```

Yes, `date_trunc` rounds `2006-10-29 00:00:00-02:30` _*up*_ to `2006-10-29 00:00:00-03:30`, and rounds the later `2006-10-28 23:01:00-03:30` _*down*_ to the earlier `2006-10-28 00:00:00-02:30`. 😢

**Comment by DaveCTurner** [comment]
We concluded in today's Search & Aggs meeting that the current behaviour is desirable and the assertion in the test is at fault. The fix is therefore to change the assertion and not do this.
